### PR TITLE
Wrangle benchmarks

### PR DESCRIPTION
**Pull request by kazimuth**
# Description of Changes

Rewrite benchmarks to allow direct comparison between:
- `spacetime_module`: Spacetime via a module
- `spacetime_raw`: Spacetime via the RelationalDB struct
- `sqlite`

And add some targeted serialization benchmarks.

### CODE REVIEW

**Comment by kazimuth** [comment]
benchmarks please

File: crates/bench/src/lib.rs [hunk 1]
```diff
@@ -1,13 +1,125 @@
...
+                .with_writer(io::stdout);
+            let env_filter_layer = tracing_subscriber::EnvFilter::from_default_env();
+            tracing_subscriber::Registry::default()
+                .with(fmt_layer)
+                .with(env_filter_layer)
+                .init();
+        });
+    }
+
+    fn sanity<DB: BenchDatabase, T: BenchTable + RandomTable>(
```
Comment (kulakowski) [inline crates/bench/src/lib.rs:45]: I'd slightly prefer `check_basic_invariants` or something that's more specific.

File: crates/bench/src/lib.rs [hunk 2]
```diff
@@ -1,13 +1,125 @@
...
+        table_style: TableStyle,
+        in_memory: bool,
+    ) -> ResultBench<()> {
+        prepare_tests();
+
+        let mut db = DB::build(in_memory, false)?;
+        let table_id = db.create_table::<T>(table_style)?;
+        assert_eq!(db.count_table(&table_id)?, 0, "tables should begin empty");
+
+        let count = 37;
```
Comment (kulakowski) [inline crates/bench/src/lib.rs:55]: Comment or constant from elsewhere if possible.

File: modules/benchmarks/src/lib.rs [hunk 3]
```diff
@@ -1,183 +1,359 @@
...
+#[spacetimedb(table)]
+#[spacetimedb(index(btree, name = "id", id))]
+#[spacetimedb(index(btree, name = "x", x))]
+#[spacetimedb(index(btree, name = "y", y))]
+pub struct MultiIndexLocation {
+    id: u32,
+    x: u64,
+    y: u64,
+}
+// ---------- / SYNCED CODE ----------
```
Comment (kulakowski) [inline modules/benchmarks/src/lib.rs:59]: Maybe `END SYNCED CODE` instead of html-style `/ SYNCED CODE`

File: modules/benchmarks/src/lib.rs [hunk 4]
```diff
@@ -1,183 +1,359 @@
...
+
+#[spacetimedb(reducer)]
+pub fn empty() {}
+
+// ---------- insert ----------
```
Comment (kulakowski) [inline modules/benchmarks/src/lib.rs:64]: One kind of documentation would be super useful for this, and that's the framework that you used to derive all of these. Suppose we add some new operation, I'd love to then be able to methodically add the "correct" set of new things here.

File: crates/bench/src/sqlite.rs [hunk 5]
```diff
@@ -1,13 +1,36 @@
...
+    AlgebraicType, AlgebraicValue, ProductType,
+};
+use std::{fmt::Write, hint::black_box};
+use tempdir::TempDir;
 
-impl BuildDb for Connection {
-    fn build(prefill: bool, fsync: bool) -> ResultBench<Self>
+pub struct SQLite {
+    db: Connection,
+    _temp_dir: TempDir,
```
Comment (kulakowski) [inline crates/bench/src/sqlite.rs:16]: The finickiness of the TempDir drop stuff makes me want a comment here, we've had enough bugs in tests and CI about it that I don't want this refactored away to a local or whatever.

File: crates/bench/src/sqlite.rs [hunk 6]
```diff
@@ -16,81 +39,200 @@ impl BuildDb for Connection {
...
+            db,
+            _temp_dir: temp_dir,
+        })
+    }
+
+    type TableId = String;
 
-        if prefill {
-            prefill_data(&mut db, Runs::Small)?;
+    fn create_table<T: BenchTable>(&mut self, table_style: crate::schemas::TableStyle) -> ResultBench<Self::TableId> {
```
Comment (kulakowski) [inline crates/bench/src/sqlite.rs:50]: Mario or someone might be a better reviewer of the sqlite stuff here than me.

File: crates/bench/src/spacetime_raw.rs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        self.db.with_auto_commit(|tx| {
+            for row in self.db.iter(tx, prepared.table_id)? {
+                black_box(row);
+            }
+            Ok(())
+        })
+    }
+
+    type PreparedFilter = PreparedFilter;
+    #[inline(never)]
```
Comment (kulakowski) [inline crates/bench/src/spacetime_raw.rs:126]: I picked this `#[inline(never)]` one arbitrarily out of all of them to comment: you should describe where and why you do this in some top level place if possible!

File: crates/bench/src/spacetime_module.rs [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+};
+
+lazy_static::lazy_static! {
+    pub static ref BENCHMARKS_MODULE: CompiledModule =
+        CompiledModule::compile("benchmarks");
+}
+
+pub struct SpacetimeModule {
+    runtime: Runtime,
+    // it's necessary for this to be dropped BEFORE the runtime.
```
Comment (kulakowski) [inline crates/bench/src/spacetime_module.rs:19]: I personally prefer this (an explicit Drop impl) to e.g. relying on a comment about relying on the drop ordering based on field order. Maybe Mazdak or Noa have a stronger style opinion about it, I'd be curious.

File: crates/bench/src/schemas.rs [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Location {
+    id: u32,
+    x: u64,
+    y: u64,
+}
+// ---------- / SYNCED CODE ----------
+
+/// Note: the first field will be used as the primary key, when using
```
Comment (kulakowski) [inline crates/bench/src/schemas.rs:26]: I can't understand how this comment relates to this trait. The first field of what?

**Review by kulakowski — APPROVED** [review]
Awesome. Inline is mostly minor comments about comments. I think this PR can land whenever, as far as I'm concerned, but I want two things.

One is the framework or schema or strategy or whatever you want to call it, by which you derived the set of benchmarks to write out. If Joshua (say) adds an Update operation that isn't just a delete+insert (say), I'd love for him to be able to read that framework and just slot in exactly the right set of benchmarks. I'd like this to be part of this PR.

Two, if there's a ways to go, I'd like other people to be able to see either in the code or in a ticket how far there is to go until this work is complete. I'm happy if there are more PRs following this one until you get there.

File: crates/bench/benches/special.rs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    spacetime_module::BENCHMARKS_MODULE,
+};
+use spacetimedb_lib::{
+    sats::{self, BuiltinValue},
+    AlgebraicValue, ProductValue,
+};
+use spacetimedb_testing::modules::start_runtime;
+
+fn criterion_benchmark(c: &mut Criterion) {
+    /*
```
[the remark this answers is not in the document]
Comment (kazimuth) [reply crates/bench/benches/special.rs:14]: These benchmarks won't change frequently and are mostly only interesting to people writing benchmarks, but I used them to make a decision at one point so left them in. I'll add a comment.

File: crates/bench/src/schemas.rs (hunk 9 above)
Comment (kazimuth) [reply crates/bench/src/schemas.rs:26]: Of the product type / value conversion of the underlying struct, I'll move it.

File: crates/bench/src/sqlite.rs (hunk 5 above)
Comment (kazimuth) [reply crates/bench/src/sqlite.rs:16]: In fact I think refactoring it to a local could cause the tempdir to be dropped, which might break things... then again, the earlier benchmarks version did that and the sqlite benches still worked. So idk.

**Comment by kazimuth** [comment]
@kulakowski, I'll address your comments and then I think this is ready for merge.

> One is the framework or schema or strategy or whatever you want to call it, by which you derived the set of benchmarks to write out. If Joshua (say) adds an Update operation that isn't just a delete+insert (say), I'd love for him to be able to read that framework and just slot in exactly the right set of benchmarks. I'd like this to be part of this PR.

Good point, I'll document this.

> Two, if there's a ways to go, I'd like other people to be able to see either in the code or in a ticket how far there is to go until this work is complete. I'm happy if there are more PRs following this one until you get there.

Yeah, I wasn't really sure where I was going while working on this which meant I had a hard time communicating what I was doing. I have some more targeted benchmark ideas which I'll stick in a followup PR.

I do have one remaining concern, which is benchmark time. Looking at the tests on this PR, it seems that running benchmarks on GitHub actions currently takes about 80 minutes. I could strip out some / all of the sqlite benchmarks since they won't change frequently. We could also disable the in-memory benchmarks on GitHub actions, which would get rid of one axis of this giant hypercube. I think both of those provide useful information, though.

File: crates/core/src/host/wasm_common/module_host_actor.rs [hunk 11]
```diff
@@ -276,6 +276,21 @@ impl<T: WasmModule> Module for WasmModuleHostActor<T> {
...
         })
     }
+
+    fn clear_table(&self, table_name: String) -> Result<(), anyhow::Error> {
+        let db = &*self.worker_database_instance.relational_db;
+        db.with_auto_commit(|tx| {
+            let tables = db.get_all_tables(tx)?;
+            for table in tables {
+                if table.table_name != table_name {
+                    continue;
```
[the remark this answers is not in the document]
Comment (kazimuth) [reply crates/core/src/host/wasm_common/module_host_actor.rs:286]: I wasn't sure if there could be multiple tables with the same name here, can there be?

**Comment by kazimuth** [comment]
Oh, also, should I rename "Person" to "IntStringInt" and "Location" to "IntIntInt" or something similar?

**Comment by kazimuth** [comment]
I have a path to fix benchmark times -- it's because everything is being run twice, including the sqlite benchmarks. I'll mess with our Github Actions benchmark script to fix that.

File: crates/bench/src/sqlite.rs [hunk 12]
```diff
@@ -16,81 +47,238 @@ impl BuildDb for Connection {
             "PRAGMA journal_mode = WAL; PRAGMA synchronous = off;"
```
Comment (mamcx) [inline crates/bench/src/sqlite.rs:47]: Is possible to impl the TODO here?

**Review by mamcx — CHANGES_REQUESTED** [review]

File: crates/bench/src/sqlite.rs (hunk 12 above)
Comment (kazimuth) [reply crates/bench/src/sqlite.rs:47]: Done!

File: crates/bench/src/sqlite.rs [hunk 13]
```diff
@@ -16,81 +47,238 @@ impl BuildDb for Connection {
...
+            }
+            _ => unimplemented!(),
+        }
+
+        commit.execute(())?;
+        Ok(())
+    }
+}
+
+const BEGIN_TRANSACTION: &str = "BEGIN DEFERRED";
```
[the remark this answers is not in the document]
Comment (kazimuth) [reply crates/bench/src/sqlite.rs:219]: The rusqlite transaction API actually just invokes these statements, but it doesn't cache them, which ends up penalizing performance significantly. So we do the same thing but cache it.

**Review by mamcx — APPROVED** [review]
LGTM

**Pull request merged**